### ISSUE TITLE
Stabilize channel state API

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#4359]
ManagedChannel.getState(), ManagedChannel.notifyWhenStateChanged(), and ConnectivityState.

**Comment by chbatey** [comment]
Would it be possible to extend this API to observe every state change? E.g. rather than a one off callback a  consumer that receives the previous and new state for each state change?

**Comment by ejona86** [comment]
Not really. It was an explicit design goal to avoid showing every state change. Basically, state changes can happen very rapidly and that would cause them to queue up, depending on the processing speed of the application.

@chbatey, what is your use-case for seeing every transition?

**Comment by abekenza** [comment]
Not sure if I am writing in correct thread, but in my use case users should be able to see current connection state like: connecting and connected

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/4569912/43801538-3e40a32a-9a48-11e8-9605-83eff0e84225.gif)

**Comment by ejona86** [comment]
@abekenza, yes, this is the correct API. You can use `ManagedChannel.getState()` and `notifyWhenStateChanged()`

**Comment by ptravers** [comment]
:wave: I am planning to use this API in a production setting so was curious as to what work is left to stabilise this API?

**Comment by ejona86** [comment]
There are no plans to change the methods and enum list. However, the specific meaning/transitions between states is a little in flux. In particular, after TRANSIENT_FAILURE we go back to IDLE (and then later CONNECTING) after a backoff period. However, that causes "lurches" of extremely poor RPC latency when things are broken, for little benefit, because we queue RPCs for IDLE/CONNECTING and broken networks can take a long time for connection attempts to fail. It is described in part in https://github.com/grpc/grpc-java/issues/6650 . But essentially, I want to make a similar change to pick_first (at which point we've clearly minorly redefined the meaning of the enum states). But there's difficulty with pick_first that needs to be discussed cross-language.

I don't expect any changes to break users of this API using it for health/status notification. But if someone is using the API and changing their RPC behavior because of what it returns (e.g., not sending RPCs if the Channel is TRANSIENT_FAILURE), then there could be surprises (it _should_ be fine, but technically it is a behavior/API change and I really don't want to argue with people if it causes minor problems). I also wouldn't be surprised if _some_ users were using the API to reverse engineer our exponential backoff state, and that would definitely break.

**Comment by larry-safran** [comment]
This was approved on June 7th for stabilization.

**Comment by larry-safran** [comment]
The 2 methods are okay, but for ConnectivityState we recently changed the behavior with pickFirst sticky TF state and some of the descriptions are not true.
Shouldn't talk about transitions in the state descriptions.

Prereq cross language definition of states
